### ISSUE TITLE
Zenject is not really thread safe

### ISSUE THREAD

**Issue by abbadryt** [modesttree/Zenject#35]
The code bellow will fail in a multi-threaded application. As you can see, there are two different locks to access the `_typeInfo` dictionary, one to `TryGetValue` and the other to `Add`.
This will raise an exception when two threads pass the first lock and both try to add the type info into the dictionary.
Locally I've fixed this by making only one big lock, encapsulating all the code from TryGetValue to Add, but I'm not sure if this is the best approach as I'm not familiar with the rest of the library.

https://github.com/svermeulen/Extenject/blob/32d175f28450530cd160f866d970af07487c822f/UnityProject/Assets/Plugins/Zenject/Source/Util/TypeAnalyzer.cs#L117-L152

**Comment by svermeulen** [comment]
Yes I see what you mean.  We can't just put the whole contents of TryGetInfo into a lock though because it recurses on itself when resolving parent values for InjectTypeInfo.  I've changed it `_typeInfo[type] = info;` instead of `_typeInfo.Add(type, info)` to avoid the exception as this seems to be the lowest risk fix.
